### DISCUSSION TITLE
How to get server IP from Httpx get response

### DISCUSSION THREAD

**Discussion by ElliotGarbus** [encode/httpx#2633] [Q&A] [answered]
How should I get the server IP address from the Httpx get response?
I have not found an answer in the docs. This stack overflow question, https://stackoverflow.com/questions/73934230/get-server-ip-from-response-of-httpx-get , provides the rather implementation dependent solution:
```python
import httpx

with httpx.Client() as hc:
    r = hc.get("https://httpbin.org/get")
    peername = r.stream._stream._httpcore_stream._stream._connection._network_stream._sock.getpeername()
    print(r.json())
    print(peername)
```
Is there a better way?

**Comment by ElliotGarbus** [comment]
Answering my own question... The trio library provides an async sockets implementation.  I will use that. 
see; https://trio.readthedocs.io/en/stable/reference-io.html#trio.socket.getaddrinfo

**Comment by lovelydinosaur** [comment, marked as answer]
> Is there a better way?

There is a better way, yes...

That answer is digging into private API, so don't expect it to be resilient.

We do provide a response extension which gives a *public API* onto this information.
It is currently only documented in the low level `httpcore` library docs, here... https://www.encode.io/httpcore/extensions/#extra-network-information
But we should also document it, since it's available through `httpx` too.

See issue https://github.com/encode/httpx/issues/2599

Adapting your example to include the server address info...

```python
import httpx

with httpx.Client() as client:
    with client.stream("GET", "https://httpbin.org/get") as response:
        network_stream = response.extensions["network_stream"]
        server_addr = network_stream.get_extra_info("server_addr")
        print(response.json())
        print(server_addr)
```

Notice that we're using `stream` here so that we've got a context where the network connection guaranteed to still available at the point we're inspecting the server address. If we used a `client.get()` then the network connection might or might not still be open *after* we've read the response.

**Reply by lovelydinosaur** [reply]
Note also that this API ought to work fine for *any* of the sync, asyncio, or trio variants.

(Although the code would look a little different for the async cases)

**Reply by robe1149** [reply]
Thank you for this!!

**Comment by ElliotGarbus** [comment]
Many thanks!
